src/java.base/share/classes/java/net/URLStreamHandler.java

```diff
@@ -259,26 +253,27 @@ protected void parseURL(URL u, String spec, int start, int limit) {
 
         // Parse the file path if any
         if (start < limit) {
+            String specStr = spec.substring(start, limit);
             if (spec.charAt(start) == '/') {
-                path = spec.substring(start, limit);
+                path = specStr;
             } else if (path != null && !path.isEmpty()) {
                 isRelPath = true;
                 int ind = path.lastIndexOf('/');
-                String separator = "";
-                if (ind == -1 && authority != null)
-                    separator = "/";
-                path = path.substring(0, ind + 1) + separator +
-                         spec.substring(start, limit);
-
+                if (ind == -1 && authority != null) {
+                    path = "/".concat(specStr);
+                } else {
+                    path = path.substring(0, ind + 1).concat(specStr);
+                }
             } else {
-                path = spec.substring(start, limit);
-                path = (authority != null) ? "/" + path : path;
+                path = (authority != null) ? "/".concat(specStr) : specStr;
             }
         } else if (queryOnly && path != null) {
             int ind = path.lastIndexOf('/');
-            if (ind < 0)
-                ind = 0;
-            path = path.substring(0, ind) + "/";
+            if (ind < 0) {
+                path = "/";
+            } else {
+                path = path.substring(0, ind).concat("/");
```

Comment:
would that be equivalent to 
```
path = path.substring(0, ind + 1);
```
given that ind = path.lastIndexOf('/') ?

Reply:
Right, fixed!